### PR TITLE
fix: transaction builder `no_input_to_cover_fees`

### PR DESCRIPTION
**Pull request by hal3e**
The function `no_spendable_input` was not checking if the `asset_id` was the base asset. Coins with other asset_ids can not be used to cover the tx fee.

### 
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: packages/fuels-core/src/types/transaction_builders.rs [hunk 1]
```diff
@@ -405,14 +406,12 @@ impl ScriptTransactionBuilder {
             .collect()
     }
 
-    fn no_spendable_input<'a, I: IntoIterator<Item = &'a FuelInput>>(inputs: I) -> bool {
+    fn no_input_to_cover_fees<'a, I: IntoIterator<Item = &'a FuelInput>>(inputs: I) -> bool {
         !inputs.into_iter().any(|i| {
             matches!(
                 i,
-                FuelInput::CoinSigned(_)
```
Comment (MujkicA) [inline packages/fuels-core/src/types/transaction_builders.rs:412]: Why are messages disregarded?

Comment (hal3e) [reply packages/fuels-core/src/types/transaction_builders.rs:412]: They should be included. Pushed a new commit.

**Review by MujkicA — APPROVED** [review]

**Review by digorithm — APPROVED** [review]

**Comment by chlenc** [comment]
Guys, when do you plan to release this fix?

File: packages/fuels-core/src/types/transaction_builders.rs [hunk 2]
```diff
@@ -405,16 +407,20 @@ impl ScriptTransactionBuilder {
...
+                | FuelInput::CoinPredicate(CoinPredicate { asset_id, .. })
+                    if *asset_id == BASE_ASSET_ID =>
+                {
+                    true
+                }
+                FuelInput::MessageCoinSigned(_) | FuelInput::MessageCoinPredicate(_) => true,
+                _ => false,
+            })
+            .not()
     }
```
Comment (segfault-magnet) [inline packages/fuels-core/src/types/transaction_builders.rs:424]: Some points:
1. `no_input_to_cover_fees` implies that any base input input will be able to cover fees which is misleading. Maybe indicate that you don't care about the amount but rather that only an input of the base asset exists.
2. The fn reads weird. I start off reading the name saying "i say whether we have no inputs to cover the fee". With that in mind I proceed to read that we're asking whether we have any input that CAN cover it. That took me a few reads until I saw the `not` at the end. I suggest a more linear flow by either inverting the function name or by introducing a temp variable:

So something like this:
```rust
    fn no_base_asset_input<'a, I: IntoIterator<Item = &'a FuelInput>>(inputs: I) -> bool {
        let has_base_asset = inputs.into_iter().any(|i| match i {
            FuelInput::CoinSigned(CoinSigned { asset_id, .. })
            | FuelInput::CoinPredicate(CoinPredicate { asset_id, .. })
                if *asset_id == BASE_ASSET_ID =>
            {
                true
            }
            FuelInput::MessageCoinSigned(_) | FuelInput::MessageCoinPredicate(_) => true,
            _ => false,
        });
        !has_base_asset
    }
```

**Review by segfault-magnet — DISMISSED** [review]

**Review by MujkicA — APPROVED** [review]

**Review by segfault-magnet — APPROVED** [review]

**Review by chlenc — APPROVED** [review]

**Review by PaulZhemanov — APPROVED** [review]

**Review by digorithm — APPROVED** [review]

**Comment by digorithm** [comment]
> Guys, when do you plan to release this fix?

It's been published just now: https://github.com/FuelLabs/fuels-rs/releases/tag/v0.55.1

**Pull request merged**